### PR TITLE
LSP Server for Java

### PR DESCRIPTION
**Pull request by jlahoda**

### CODE REVIEW

**Comment by geertjanw** [comment]
Wow, so much cool stuff here.

**Comment by mickaelistria** [comment]
Is the goal implementing an alternative to JDT-LS ?
Is it described somewhere why an alternative to JDT-LS is desired in first place and what would be the differentiators between this and JDT-LS? (cc @fbricon)

**Comment by neilcsmith-net** [comment]
@mickaelistria surely the obvious answer is javac?!

**Comment by mickaelistria** [comment]
If standard JDK is the key feature, then how is it different from https://github.com/georgewfraser/java-language-server ?
(Note that I'm really not trying to sound negative, I'm asking those questions to make sure I understand what cases are going to be better covered by this LS than by existing ones)

**Comment by jlahoda** [comment]
So, I think this is a good question.

This project is maintaining editor features that support developing in the Java language (and we could ask why, although that is probably out of the scope of this PR). So adding a fairly small layer that exports a subset of these features through the LSP does not feel to be very off. It is simply providing a way for those that prefer a different UI to use Java editor features this project already has. Not that this PR would be quite there.

The NB Java editor has an ability to use JDK's javac for parsing, so it can in some cases support editing of files with Java language features that are newer than the editor itself. And if a new NB feature is added supporting such a new language feature, the LSP server may use it right away. Yes, other server will get that as well, but as long as we maintain the stuff here,  why not allow access to it using the LSP as well? This is niche, of course.

**Comment by neilcsmith-net** [comment]
> If standard JDK is the key feature, then how is it different from https://github.com/georgewfraser/java-language-server ?

Until recently that project was also using parts of (pre-Apache) NetBeans' - eg. `nb-javac`.  It also appears at present to have a [license issue](https://github.com/georgewfraser/java-language-server/blob/master/src/main/java/org/javacs/ReusableCompiler.java).

Personally I'd prefer there was an LS as part of the OpenJDK tools.  But having it here seems next best step for an LS backed by javac and building on an editing infrastructure that's widely used.

**Comment by geertjanw** [comment]
Should this be merged?

**Comment by jlahoda** [comment]
I am definitely +1 on merging this :-).

File: java/java.lsp.server/vscode/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+This is an extension for VS Code. Based on "lsp-sample" from:
```
Comment (junichi11) [inline java/java.lsp.server/vscode/README.md]: Missing the license header (Travis CI fails)

**Comment by geertjanw** [comment]
OK, merging.

**Pull request merged**